owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/status/GetRemoteCapabilitiesOperation.kt

```diff
@@ -0,0 +1,323 @@
+ *   Copyright (C) 2019 ownCloud GmbH.
+ *
+ *   Permission is hereby granted, free of charge, to any person obtaining a copy
+ *   of this software and associated documentation files (the "Software"), to deal
+ *   in the Software without restriction, including without limitation the rights
+ *   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *   copies of the Software, and to permit persons to whom the Software is
+ *   furnished to do so, subject to the following conditions:
+ *
+ *   The above copyright notice and this permission notice shall be included in
+ *   all copies or substantial portions of the Software.
+ *
+ *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ *   BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ *   ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ *   CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *   THE SOFTWARE.
+ *
+ */
+
+package com.owncloud.android.lib.resources.status
+
+import com.owncloud.android.lib.common.OwnCloudClient
+import com.owncloud.android.lib.common.http.HttpConstants
+import com.owncloud.android.lib.common.http.methods.nonwebdav.GetMethod
+import com.owncloud.android.lib.common.operations.RemoteOperation
+import com.owncloud.android.lib.common.operations.RemoteOperationResult
+import com.owncloud.android.lib.common.operations.RemoteOperationResult.ResultCode.OK
+import com.owncloud.android.lib.common.utils.Log_OC
+import org.json.JSONObject
+import java.net.URL
+
+/**
+ * Get the Capabilities from the server
+ * Save in Result.getData in a RemoteCapability object
+ *
+ * @author masensio
+ * @author David González Verdugo
+ */
+/**
+ * Constructor
+ */
+class GetRemoteCapabilitiesOperation : RemoteOperation<RemoteCapability>() {
+
+    override fun run(client: OwnCloudClient): RemoteOperationResult<RemoteCapability> {
+        var result: RemoteOperationResult<RemoteCapability>
+
+        try {
+            val requestUri = client.baseUri
+            val uriBuilder = requestUri.buildUpon()
+            uriBuilder.appendEncodedPath(OCS_ROUTE)    // avoid starting "/" in this method
+            uriBuilder.appendQueryParameter(PARAM_FORMAT, VALUE_FORMAT)
+
+            val getMethod = GetMethod(URL(uriBuilder.build().toString()))
+
+            getMethod.addRequestHeader(RemoteOperation.OCS_API_HEADER, RemoteOperation.OCS_API_HEADER_VALUE)
+
+            val status = client.executeHttpMethod(getMethod)
+
+            val response = getMethod.responseBodyAsString
+            if (isSuccess(status)) {
+                Log_OC.d(TAG, "Successful response: " + response!!)
+
+                // Parse the response
+                val respJSON = JSONObject(response)
```

Comment:
I don't know why you don't use Gson, Moshi or Jackson for reducing this code...

Reply:
We are already using Gson in the app but not in the library, I will include this as an enhancement task in the new arch epic issue and try to face it within this PR, thanks for the suggestion

Reply:
Well, the capabilities json is not an easy one, it includes many nodes, some on them in different sublevels so the parsing to Java objects is not trivial at all. I'd say that using Gson here is not worth it since it will require the same or even more code than the current one.